IO/UnbufferedStreamReader.cs

```diff
@@ -0,0 +1,187 @@
+        /// <returns>The character read or default <see cref="char"/> if at the end of the stream.</returns>
+        public char Read()
+        {
+            return ReadNextCharacter();
+        }
+
+        /// <summary>
+        /// Reads the next line from the stream.
+        /// </summary>
+        /// <returns>The line read or <see cref="string.Empty"/> if already at the end of the stream.</returns>
+        public string ReadLine()
+        {
+            if (!m_IsPreambleHandled)
+            {
+                CheckPreambleMatchesEncoding();
+            }
+
+            while (true)
+            {
+                char character = ReadNextCharacter();
+
+                if (character == default || character == '\r' || character == '\n')
+                {
+                    return RenderAndClearStringBuilder();
+                }
+
+                m_StringBuilder.Append(character);
+            }
+        }
+
+        /// <summary>
+        /// Reads from the current stream position to the end of the stream.
+        /// </summary>
+        /// <returns>The content read or <see cref="string.Empty"/> if at the end of the stream.</returns>
+        public string ReadToEnd()
+        {
+            if (!m_IsPreambleHandled)
+            {
+                CheckPreambleMatchesEncoding();
+            }
+
+            while (true)
+            {
+                char character = ReadNextCharacter();
+
+                if (character == default)
+                {
+                    return RenderAndClearStringBuilder();
+                }
+
+                m_StringBuilder.Append(character);
+            }
+        }
+
+        private void CheckPreambleMatchesEncoding()
+        {
+            if (m_IsPreambleHandled)
+            {
+                return;
+            }
+
+            byte[] preamble = CurrentEncoding.GetPreamble();
+            for (int i = 0; i < preamble.Length; i++)
+            {
+                byte readByte = (byte)BaseStream.ReadByte();
+                if ( readByte != preamble[i] )
+                {
+                    throw new InvalidDataException($"Preamble does not match encoding. Invalid byte index: {i}, value: {readByte}, expected: {preamble[i]}");
+                }
+            }
+
+            m_IsPreambleHandled = true;
+        }
+
+        // Returns default char if at end of the stream
+        private char ReadNextCharacter()
+        {
+            Debug.Assert(m_IsPreambleHandled);
+
+            int decodedCharCount;
+            do
+            {
+                m_SingleByteBuffer[0] = (byte)BaseStream.ReadByte();
+                if (m_SingleByteBuffer[0] < 0)
+                {
+                    return default;
+                }
+
+                decodedCharCount = m_Decoder.GetChars(m_SingleByteBuffer, 0, 1, m_charBuffer, 0);
+            } while (decodedCharCount < 1);
+
+            return m_charBuffer[0];
+        }
+
+        private string RenderAndClearStringBuilder()
```

Comment:
Seems odd to call it `Render`?  Maybe `ResolveAndClearStringBuilder`? `ExtractAndClear`?

Reply:
I mean, you are rendering a string from the builder's component parts but sure I've rename it to resolve.

Reply:
Not to the screen man.... not to the screen.